ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/QueryFilterNode.java

```diff
@@ -447,10 +441,44 @@ public Void visitComparisonExpression(
         final Object key = resolveKey(other, col.get(), metaStore, ksqlConfig, node);
         keyContents[col.get().index()] = key;
         seenKeys.set(col.get().index());
+        type = node.getType();
+        if (other instanceof StringLiteral) {
+          isString = true;
+        }
       }
       return null;
     }
 
+    public LookupConstraint getLookupConstraint(final Optional<WindowBounds> windowBounds) {
+      if (seenKeys.isEmpty()) {
+        return new NonKeyConstraint();
+      }
+
+      if (type == Type.EQUAL) {
+        return new KeyConstraint(ConstraintOperator.EQUAL,
+          GenericKey.fromList(Arrays.asList(keyContents)), windowBounds);
+      } else if (isString) {
```

Comment:
Yea, for non string types we currently fallback to NonKeyContraint (which leads to a tablescan) because we can't guarantee the order returned by the state store is going to be correct for non-string types.